File: signal-layer-server/app/schemas/signal.py
from typing import Literal, Optional, Union, Annotated
from pydantic import AliasChoices, BaseModel, Field, Discriminator


# ============================================================
# Trade Parameters
# ============================================================

class TradeParams(BaseModel):
    # 止损
    stop_loss_type: Literal["atr", "fixed_pct", "swing_low"] = "atr"
    stop_loss_value: float = 2.0
    # 止盈
    take_profit_type: Literal["atr", "fixed_pct", "rr_ratio"] = "rr_ratio"
    take_profit_value: float = 2.0
    # 仓位
    position_type: Literal["fixed_pct", "kelly"] = "fixed_pct"
    position_value: float = 20.0
    # 条件式出场（复用条件组结构，引用指标/价格/缠论做判断）
    exit_conditions: list["ConditionGroupSchema"] = Field(default_factory=list)
    exit_logic: Literal["AND", "OR"] = "AND"


# ============================================================
# Condition value types
# ============================================================

class PriceValue(BaseModel):
    source: Literal["price"]
    field: Literal["open", "high", "low", "close", "volume"]


class IndicatorValue(BaseModel):
    source: Literal["indicator"]
    indicator_type: str
    params: dict[str, float] = Field(default_factory=dict)
    field: str = "value"


class ChanValue(BaseModel):
    source: Literal["chan"]
    element: Literal["fenxing", "bi", "zhongshu", "divergence", "buySellPoint"]
    property: Optional[str] = None


class ConstantValue(BaseModel):
    source: Literal["constant"]
    value: float


class TimeframeValue(BaseModel):
    source: Literal["timeframe"]
    timeframe_id: str
    inner: "ConditionValueUnion"


ConditionValueUnion = Union[
    PriceValue, IndicatorValue, ChanValue, ConstantValue, TimeframeValue
]


class ConditionSchema(BaseModel):
    id: str
    name: str
    left: ConditionValueUnion = Field(discriminator="source")
    operator: Literal[
        "gt", "gte", "lt", "lte", "eq", "crossAbove", "crossBelow",
        "rising", "falling", "turnDown", "turnUp", "between"
    ]
    right: ConditionValueUnion = Field(discriminator="source")
    right2: Optional[ConditionValueUnion] = Field(
        default=None,
        validation_alias=AliasChoices("right2", "right_2"),
    )
    timeframe_id: Optional[str] = None
    enabled: bool = True


class ConditionGroupSchema(BaseModel):
    id: str
    name: str = ""
    conditions: list[ConditionSchema]


class ConditionTemplateSchema(BaseModel):
    id: str
    name: str
    logic: Literal["AND", "OR"]
    condition_groups: list[ConditionGroupSchema]
    primary_tf: str = "15m"
    secondary_tfs: list[str] = Field(default_factory=list)
    enabled: bool = True
    trade_params: Optional[TradeParams] = None


class EvaluateRequest(BaseModel):
    symbol: str
    template: ConditionTemplateSchema


class ConditionEval(BaseModel):
    condition_id: str
    satisfied: bool
    left_value: float = 0.0
    right_value: float = 0.0
    diff_percent: float = 0.0


class GroupEval(BaseModel):
    group_id: str
    evaluations: list[ConditionEval]
    satisfied: bool


class SignalResult(BaseModel):
    template_id: str
    state: Literal["evaluating", "partial", "ready"]
    groups: list[GroupEval]
    is_ready: bool
    progress_percent: int


class BacktestRequest(BaseModel):
    symbol: str
    template: ConditionTemplateSchema
    kline_limit: int = Field(default=500, ge=50, le=2000)


class TradeRecord(BaseModel):
    entry_time: int
    exit_time: int
    side: str
    entry_price: float
    exit_price: float
    pnl_pct: float
    exit_reason: str


class BacktestResult(BaseModel):
    template_id: str
    symbol: str
    timeframe: str
    total_trades: int
    win_trades: int
    win_rate: float
    total_return: float
    avg_return: float
    max_drawdown: float
    profit_factor: float
    trades: list[TradeRecord]

File: signal-layer/src/api/signal.ts
import { api } from './client'
import type {
  ConditionTemplate,
  Condition,
  ConditionValue,
  TemplateSignal,
  GroupEvaluation,
  ConditionEvaluation,
  SignalState,
} from '../core/types.ts'

/* ========== Backend snake_case interfaces ========== */

interface ApiConditionValue {
  source: string
  field?: string
  indicator_type?: string
  params?: Record<string, number>
  element?: string
  property?: string
  value?: number
  timeframe_id?: string
  inner?: ApiConditionValue
}

interface ApiCondition {
  id: string
  name: string
  left: ApiConditionValue
  operator: string
  right: ApiConditionValue
  right2?: ApiConditionValue
  timeframe_id?: string
  enabled: boolean
}

interface ApiConditionGroup {
  id: string
  name?: string
  logic?: 'AND' | 'OR'
  conditions: ApiCondition[]
}

export interface ApiTemplate {
  id: string
  name: string
  logic: string
  condition_groups: ApiConditionGroup[]
  primary_tf: string
  secondary_tfs: string[]
  created_at: number
  updated_at: number
  enabled: boolean
  trade_params?: {
    stop_loss_type: string
    stop_loss_value: number
    take_profit_type: string
    take_profit_value: number
    exit_conditions: ApiConditionGroup[]
    exit_logic: string
  }
}

interface ApiConditionEvaluation {
  condition_id: string
  satisfied: boolean
  left_value: number | null
  right_value: number | null
  diff_percent: number
}

interface ApiGroupEvaluation {
  group_id: string
  evaluations: ApiConditionEvaluation[]
  satisfied: boolean
}

interface ApiTradeRecord {
  entry_time: number
  exit_time: number
  entry_price: number
  exit_price: number
  pnl_pct: number
  exit_reason: string
}

interface ApiBacktestResponse {
  template_id: string
  symbol: string
  timeframe: string
  total_trades: number
  win_trades: number
  win_rate: number
  total_return: number
  avg_return: number
  max_drawdown: number
  profit_factor: number
  payoff_ratio: number
  suggested_position: number
  trades: ApiTradeRecord[]
}

export interface TradeRecord {
  entryTime: number
  exitTime: number
  entryPrice: number
  exitPrice: number
  pnlPct: number
  exitReason: string
}

export interface BacktestResult {
  templateId: string
  symbol: string
  timeframe: string
  totalTrades: number
  winTrades: number
  winRate: number
  totalReturn: number
  avgReturn: number
  maxDrawdown: number
  profitFactor: number
  payoffRatio: number
  suggestedPosition: number
  trades: TradeRecord[]
}

interface ApiEvaluateResponse {
  template_id: string
  state: string
  groups: ApiGroupEvaluation[]
  is_ready: boolean
  progress_percent: number
}

/* ========== camelCase ← → snake_case conversion ========== */

function conditionValueToSnake(v: ConditionValue): ApiConditionValue {
  if (v.source === 'indicator') {
    return {
      source: 'indicator',
      indicator_type: v.indicatorType,
      params: v.params,
      field: v.field,
    }
  }
  if (v.source === 'chan') {
    return {
      source: 'chan',
      element: v.element,
      property: v.property,
    }
  }
  if (v.source === 'timeframe') {
    return {
      source: 'timeframe',
      timeframe_id: v.timeframeId,
      inner: conditionValueToSnake(v.inner),
    }
  }
  // price or constant — structure already matches
  return v as unknown as ApiConditionValue
}

function conditionToSnake(c: Condition): ApiCondition {
  const result: ApiCondition = {
    id: c.id,
    name: c.name,
    left: conditionValueToSnake(c.left),
    operator: c.operator,
    right: conditionValueToSnake(c.right),
    enabled: c.enabled,
  }
  if (c.right2) result.right2 = conditionValueToSnake(c.right2)
  if (c.timeframeId) result.timeframe_id = c.timeframeId
  return result
}

export function templateToSnake(t: ConditionTemplate): ApiTemplate {
  const result: ApiTemplate = {
    id: t.id,
    name: t.name,
    logic: t.logic,
    condition_groups: t.conditionGroups.map((g) => ({
      id: g.id,
      name: g.name,
      logic: g.logic ?? 'AND',
      conditions: g.conditions.map(conditionToSnake),
    })),
    primary_tf: t.primaryTimeframeId,
    secondary_tfs: t.secondaryTimeframeIds,
    created_at: t.createdAt,
    updated_at: t.updatedAt,
    enabled: t.enabled,
  }
  if (t.tradeParams) {
    result.trade_params = {
      stop_loss_type: t.tradeParams.stopLossType,
      stop_loss_value: t.tradeParams.stopLossValue,
      take_profit_type: t.tradeParams.takeProfitType,
      take_profit_value: t.tradeParams.takeProfitValue,
      exit_conditions: t.tradeParams.exitConditions.map((g) => ({
        id: g.id,
        name: g.name,
        logic: g.logic ?? 'AND',
        conditions: g.conditions.map(conditionToSnake),
      })),
      exit_logic: t.tradeParams.exitLogic,
    }
  }
  return result
}

function evaluationFromSnake(e: ApiConditionEvaluation): ConditionEvaluation {
  return {
    conditionId: e.condition_id,
    satisfied: e.satisfied,
    leftValue: e.left_value,
    rightValue: e.right_value,
    diffPercent: e.diff_percent,
  }
}

function groupEvalFromSnake(g: ApiGroupEvaluation): GroupEvaluation {
  return {
    groupId: g.group_id,
    evaluations: g.evaluations.map(evaluationFromSnake),
    satisfied: g.satisfied,
  }
}

/* ========== Public API ========== */

/**
 * Evaluate a signal template against a symbol.
 * POST /api/v1/signal/evaluate
 */
export async function evaluateSignal(
  symbol: string,
  template: ConditionTemplate,
): Promise<TemplateSignal> {
  console.log('[SL:API] POST /signal/evaluate', { symbol, template: template.name })
  const raw = await api.post<ApiEvaluateResponse>('/signal/evaluate', {
    symbol,
    template: templateToSnake(template),
  })
  console.log('[SL:API] POST /signal/evaluate -> OK', { state: raw.state })
  return {
    templateId: raw.template_id,
    state: raw.state as SignalState,
    groups: raw.groups.map(groupEvalFromSnake),
    isReady: raw.is_ready,
    updatedAt: Date.now(),
    progressPercent: raw.progress_percent,
  }
}


export async function runBacktest(
  symbol: string,
  template: ConditionTemplate,
  klineLimit = 500,
): Promise<BacktestResult> {
  console.log('[SL:API] POST /signal/backtest', { symbol, template: template.name })
  const raw = await api.post<ApiBacktestResponse>('/signal/backtest', {
    symbol,
    template: templateToSnake(template),
    kline_limit: klineLimit,
  })
  console.log('[SL:API] POST /signal/backtest -> OK', { trades: raw.total_trades })
  return {
    templateId: raw.template_id,
    symbol: raw.symbol,
    timeframe: raw.timeframe,
    totalTrades: raw.total_trades,
    winTrades: raw.win_trades,
    winRate: raw.win_rate,
    totalReturn: raw.total_return,
    avgReturn: raw.avg_return,
    maxDrawdown: raw.max_drawdown,
    profitFactor: raw.profit_factor,
    payoffRatio: raw.payoff_ratio,
    suggestedPosition: raw.suggested_position,
    trades: raw.trades.map(t => ({
      entryTime: t.entry_time,
      exitTime: t.exit_time,
      entryPrice: t.entry_price,
      exitPrice: t.exit_price,
      pnlPct: t.pnl_pct,
      exitReason: t.exit_reason,
    })),
  }
}
